### ISSUE TITLE
Check newrat-implementation and potentially port Michel's changes

### ISSUE THREAD

**Issue by JohannesPfeifer** [DynareTeam/dynare#1211]
During his first try of removing `objective_function_penalty_base` @MichelJuillard made various changes to `newrat` that were reverted, but potentially still need to be ported to `master`. But I am not entirely sure what these commits do. @rattoma Could you maybe have a look.

1. In the reverted commits https://github.com/DynareTeam/dynare/commit/fe077fc7103658aab7becba95167aa9d240b07f1 and https://github.com/DynareTeam/dynare/commit/f48a026d89828e9e322333263f56b733a719eb34 the initialization of `mr_hessian` was removed.
2. In https://github.com/DynareTeam/dynare/commit/5ade8d7c6fc71cc4ee4ffbb27dad2ed6b4cf8ccd the `newratflag`was passed to the optimizer, but this was reverted in https://github.com/DynareTeam/dynare/commit/d4cf3576e468202df9d1582a045dfcfd74e2e7db . What is the correct setting here?
3. In https://github.com/DynareTeam/dynare/commit/5ade8d7c6fc71cc4ee4ffbb27dad2ed6b4cf8ccd the part 
```
if analytic_derivation,
        hhx=hh;
else
        hhx = reshape(dum,nx,nx);
end
```
was replaced by
` hhx=hh;`
But when I tried this, the program crashed in tests.

**Comment by JohannesPfeifer** [comment]
This is the part of @MichelJuillard's commits not included in https://github.com/DynareTeam/dynare/pull/1212

**Comment by rattoma** [comment]
@johannes, I think remember those reverted commits.
1) persistent `h1` and `htol` are needed to make computation quicker: re-setting them every call, makes the adaptive step size longer. If we want to get rid of persistent, we should pass them as input/output args. 
2) the correct setting is to hard-code 0. The flag is used in `dynare_estimation_1`, because it also  depends on the value of `kalman_algo`. this also avoids possibly useless calls to mr_hessian in newrat.
3) I think this was a bug? hh is only computed with analytic derivatives I think.

**Comment by johannes** [comment]
Hi @rattoma, please write @JohannesPfeifer to tag him, not @johannes, which tags and subscribes me.

**Comment by JohannesPfeifer** [comment]
@MichelJuillard Was there a particular reason you wanted to replace the persistent variables?
@rattoma Regarding 3: is there anything we need to do here to close the ticket? If `hh` is only computed with analytic derivatives, we cannot delete the if-clause as in the commit.

**Comment by MichelJuillard** [comment]
@JohannesPfeifer I believe that persistent variables are as expensive as global variables. They also make for complicated code because of their initialization

**Comment by JohannesPfeifer** [comment]
@MichelJuillard Ok, but that is a general design choice and not something we have to change immediately, I guess.

**Comment by MichelJuillard** [comment]
It would be better to make at once all changes to the function calling parameters

**Comment by JohannesPfeifer** [comment]
@MichelJuillard Ok. I did the change in eef7e3c6062cbfadcacbce4441ef3c84e4fd24fd of #1212

**Comment by rattoma** [comment]
@JohannesPfeifer thanks a lot. I am sending you a patch for your branch, when the Hessian option of newrat is not to default (i.e. not 1).
By the way, I was wandering whether we should always check if params are at the bounds, irrespective to the error in `chol(hh)` [see the test fs2000_5.mod, where newrat gets to the lower bound for `rho`], so snip this check outside the `try-catch` block in `dynare_estimation_1`

**Comment by JohannesPfeifer** [comment]
@rattoma Thanks a lot. I updated the branch

**Comment by JohannesPfeifer** [comment]
Closed by#1212. Part 3 above was a bug introduced by Michel
